motif/motif.go

```diff
@@ -0,0 +1,181 @@
+}
+
+// WritePositionMatrix writes an individual PositionMatrix struct to a fileio.EasyWriter.
+func WritePositionMatrix(file *fileio.EasyWriter, m PositionMatrix) {
+	var err error
+	_, err = fmt.Fprintf(file, ">%s\t%s\n%s", m.Id, m.Name, matToString(m.Mat))
+	exception.PanicOnErr(err)
+}
+
+//matToString is a helper function of matToString that converts a Pfm/Ppm matrix to a string.
+func matToString(mat [][]float64) string {
+	var answer string = "A\t[\t"
+	if len(mat) != 4 {
+		log.Fatalf("Error: Input PFM must have 4 rows, one for each nucleotide.")
+	}
+	for i := range mat[0] {
+		answer = answer + fmt.Sprintf("\t%g", mat[0][i])
+	}
+	answer = answer + "\t]\nC [ "
+	for i := range mat[1] {
+		answer = answer + fmt.Sprintf("\t%g", mat[1][i])
+	}
+	answer = answer + "\t]\nG [ "
+	for i := range mat[2] {
+		answer = answer + fmt.Sprintf("\t%g", mat[2][i])
+	}
+	answer = answer + "\t]\nT [ "
+	for i := range mat[3] {
+		answer = answer + fmt.Sprintf("\t%g", mat[3][i])
+	}
+	answer = answer + "\t]\n"
+	return answer
+}
+
+//StringToPositionMatrixType parses a jaspar.PositionMatrixType type from an input string.
+func StringToPositionMatrixType(s string) PositionMatrixType {
+	switch s {
+	case "Frequency":
+		return Frequency
+	case "Probability":
+		return Probability
+	case "Weight":
+		return Weight
+	default:
+		log.Fatalf("Error: expected %s to be a position matrix type that is either 'Frequency', 'Probability', or 'Weight'.\n", s)
+		return None
+	}
+}
+
+//Read parses a slice of PositionMatrix structs from an input file in JASPAR format.
+func Read(filename string, Type string) []PositionMatrix {
+	var err error
+	var curr PositionMatrix
+	var answer []PositionMatrix
+	var doneReading bool
+	usedMotifIds := make(map[string]bool)
+
+	file := fileio.EasyOpen(filename)
+
+	for curr, doneReading = NextPositionMatrix(file, StringToPositionMatrixType(Type)); !doneReading; curr, doneReading = NextPositionMatrix(file, StringToPositionMatrixType(Type)) {
+		if usedMotifIds[curr.Id] {
+			log.Fatalf("Error: %s is used as the ID for multiple records. IDs must be unique.", curr.Id)
+		} else {
+			usedMotifIds[curr.Id] = true
+			answer = append(answer, curr)
+		}
+	}
+	err = file.Close()
+	exception.PanicOnErr(err)
+	return answer
+}
+
+// NextPositionMatrix reads and parses a single PositionMatrix record from an input EasyReader. Returns true when the file is fully read.
+func NextPositionMatrix(file *fileio.EasyReader, t PositionMatrixType) (PositionMatrix, bool) {
+	var header string
+	var fields []string
+	var motifLen int
+	var answer PositionMatrix
+
+	line1, done1 := fileio.EasyNextRealLine(file)
+	line2, done2 := fileio.EasyNextRealLine(file)
+	line3, done3 := fileio.EasyNextRealLine(file)
+	line4, done4 := fileio.EasyNextRealLine(file)
+	line5, done5 := fileio.EasyNextRealLine(file)
+
+	if done1 {
+		return PositionMatrix{}, true
+	}
+	if done2 || done3 || done4 || done5 {
+		log.Fatalf("Error: There is an empty line in this Pfm record.")
```

Comment:
Maybe the error message should say something like the number of non-comment lines in the file should be a multiple of 5?